File: daily/python/Nov-17-balanced-binary-tree.py
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return (self.Height(root) >= 0)
    
    def Height(self, root):
        if root is None: return 0
        leftH, rightH = self.Height(root.left), self.Height(root.right)
        if leftH < 0 or rightH < 0 or abs(leftH - rightH > 1): return -1
        return max(leftH, rightH) + 1

File: daily/cpp/Nov-17-balanced-binary-tree.cpp
class Solution {
 public:
  bool isBalanced(TreeNode* root) {
    if (root == nullptr) return true;
    if (height(root) == -1) return false;
    return true;
  }

 private:
  int height(TreeNode* root) {
    if (root == nullptr) return 0;
    int leftH = height(root->left);
    int rightH = height(root->right);
    if (leftH == -1 || rightH == -1 || abs(leftH - rightH) > 1) return -1;
    return max(leftH, rightH) + 1;
  }
  // there is recursive helper function, time complexity is O(n)
}